marconi/src/Marconi/Index/MintBurn.hs

```diff
@@ -0,0 +1,310 @@
+sqliteSelectByTxIdPolicyId sqlCon (conditions, params) txId policyId =
+  fromRows <$> SQL.queryNamed sqlCon query ([":txId" := txId, ":policyId" := policyId] <> params)
+  where query =
+          " SELECT slotNo, blockHeaderHash, txId, policyId                    \
+          \      , assetName, quantity, redeemerIx, redeemerData              \
+          \   FROM minting_policy_events                                      \
+          \  WHERE txId = :txId AND policyId = :policyId " <> conditions <> " \
+          \  ORDER BY slotNo, txId                                            "
+
+sqliteSelectAll :: SQL.Connection -> (SQL.Query, [NamedParam]) -> IO [TxMintEvent]
+sqliteSelectAll sqlCon (conditions, params) = fromRows <$> SQL.queryNamed sqlCon query params
+  where
+    whereClause = if conditions == "" then "" else " WHERE " <> conditions <> " "
+    query =
+      "   SELECT slotNo, blockHeaderHash, txId, policyId       \
+      \        , assetName, quantity, redeemerIx, redeemerData \
+      \     FROM minting_policy_events                         \
+      \     " <> whereClause <> "                              \
+      \ ORDER BY slotNo, txId                                  "
+
+intervalToWhereClause :: RI.QueryInterval C.ChainPoint -> (SQL.Query, [NamedParam])
+intervalToWhereClause qi = case qi of
+  RI.QEverything -> ("", [])
+  RI.QInterval from to
+    | Just from' <- slotMaybe from, Just to' <- slotMaybe to -> ("slotNo BETWEEN :fromSlot AND :toSlot" , [":fromSlot" := from', ":toSlot" := to'])
+    | Just to' <- slotMaybe to -> ("slotNo <= :toSlot" , [":toSlot" := to'])
+    | otherwise -> ("FALSE", [])
+    where
+      slotMaybe :: C.ChainPoint -> Maybe C.SlotNo
+      slotMaybe = \case
+        C.ChainPoint slotNo _ -> Just slotNo
+        C.ChainPointAtGenesis -> Nothing
+
+groupBySlotNo :: [TxMintEvent] -> [TxMintEvent]
+groupBySlotNo events = events
+  & sort
+  & groupBy (\e1 e2 -> txMintEventSlotNo e1 == txMintEventSlotNo e2 && txMintEventBlockHeaderHash e1 == txMintEventBlockHeaderHash e2)
+  & concatMap (\case e : es -> [ TxMintEvent (txMintEventSlotNo e) (txMintEventBlockHeaderHash e) $ txMintEventTxAssets =<< (e :| es) ]
+                     _      -> [])
+
+-- * Indexer
+
+data MintBurnHandle = MintBurnHandle
+  { sqlConnection :: SQL.Connection
+  , securityParam :: Int
+  }
+
+type MintBurnIndexer = RI.State MintBurnHandle
+
+type instance RI.StorablePoint MintBurnHandle = C.ChainPoint
+
+type instance RI.StorableMonad MintBurnHandle = IO
+
+newtype instance RI.StorableEvent MintBurnHandle
+  = MintBurnEvent TxMintEvent
+
+data instance RI.StorableQuery MintBurnHandle
+  = ByTxIdAndPolicyId C.TxId C.PolicyId
+  | Everything
+
+data instance RI.StorableResult MintBurnHandle
+  = MintBurnResult [Row]
+
+instance RI.Queryable MintBurnHandle where
+  queryStorage queryInterval memoryEvents (MintBurnHandle sqlCon _k) query = case query of
```

Comment:
You are not getting the proper data here, see the previous comment.

You need to include all the data persisted from the point you found to the bottom of the event stack. The way that the state is computed is by running a fold over the in-memory + persisted data sets, so you will need all the data from the persistence store.

Then you stitch the persisted events and the in-memory events together and you run them through the `filterWithQueryInterval` function that should return the final result.

For reference, here how it is done in the test indexer implementation: https://github.com/input-output-hk/plutus-apps/blob/d9eebba0e5b7f3f0d955e9755c7a031eb4de1303/rewindable-index/test/RewindableIndex/Spec/Sqlite.hs#L192-L206

Reply:
Right, but given that my isn't a fold, but a map, then why do I need to load all the events from the database?